### ISSUE TITLE
Rate limiting with excluded IPs doesn't use request.RemoteAddr

### ISSUE THREAD

**Issue by manveru** [traefik/traefik#8052]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

Tried to set a ratelimit that uses the IPStrategy and ues the remote IP address, since there is no `X-Forwarded-For` header in our setup (Traefik is listening to requests directly).

### What did you expect to see?

Exhausing the rate limit from one IP should not affect the limit of any other IP, and `123.123.123.123` should be excluded from the limit entirely.

### What did you see instead?

When the `ratelimit.sourcecriterion.ipstrategy.excludedips` is set, Traefik only considers the value of `X-Forwarded-For`
It took a bit of debugging and adding log statements all over the place, but it turns out that the `source` at https://github.com/traefik/traefik/blob/v2.4.7/pkg/middlewares/ratelimiter/rate_limiter.go#L108 is just an empty string as soon as you set the excluded IPs. That means every request shares the same bucket.

The cause for this is most likely found at these places:

https://github.com/traefik/traefik/blob/v2.4.7/pkg/config/dynamic/middlewares.go#L262
https://github.com/traefik/traefik/blob/v2.4.7/pkg/ip/strategy.go#L58

There are two tests for the excluded IP behavior at https://github.com/traefik/traefik/blob/v2.4.7/pkg/ip/strategy_test.go but they both only test by setting the header.

### Output of `traefik version`:

```
Version:      2.4.7
Codename:     cheddar
Go version:   go1.16.2
Built:        I don't remember exactly
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

Following configs were set via Consul:

```
traefik.enable=true
traefik.http.routers.testing.rule=Host(`example.com`)
traefik.http.routers.testing.entrypoints=https
traefik.http.routers.testing.tls=true
traefik.http.routers.testing.middlewares=testing-ratelimit@consulcatalog
traefik.http.middlewares.testing-ratelimit.ratelimit.average=100
traefik.http.middlewares.testing-ratelimit.ratelimit.burst=250
traefik.http.middlewares.testing-ratelimit.ratelimit.period=1m
traefik.http.middlewares.testing-ratelimit.ratelimit.sourcecriterion.ipstrategy.excludedips=123.123.123.123
```

**Comment by mpl** [comment]
I think I understand what you're looking at, and I think I agree with your analysis, however I am not sure I get what you're tying to achieve.
If you want the ratelimit matcher to user the remote address (so that each incoming request from a different address gets its own bucket), then why don't you use an empty ratelimit.sourcecriterion (which defaults to using the remote address) in your config?

**Comment by manveru** [comment]
That's because I would like to exclude specific IPs from the rate limiting to allow stress testing behind the proxy.

**Comment by mpl** [comment]
@manveru If I understood the code correctly, even if we fixed the CheckerStrategy algorithm to take into account the RemoteAddress (which maybe we should, and I started working on that), I don't think it would fix your problem, as I don't think the excludedips strategy would do what I think you want (arguably, because this excludedips option is badly named).

The way the rate limiter works is: all the IPs in the excludedips pool will be considered as the same source (the "" one), and a common rate limit will be applied to all of them. But it is not meant to simply exclude them from any rate limiting considerations.

So with the configuration you showed the effect would be that: 1) any request that does not come from 123.123.123.123 gets into a different bucket (each unique IP gets its own bucket), i.e. anyone on the internet gets their own ratelimiting at 100req/m, 2) all requests matching the IPs in the pool get to the bucket for the "" (empty string), which also gets its ratelimiting at 100req/m. Since there's only one IP in the pool, it is in effect useless, and you could just as well remove the excludedips criterion.

If what you want to do is have a rate limit in general for most IPs, but have 123.123.123.123 completely escape the rate limiter, then I think what you want is two routers: one exclusively for 123.123.123.123, with no rate limit middleware applied to it, and another router for all the other requests, with a rate limit middleware applied to it, with whatever parameters fit your case.

**Comment by manveru** [comment]
Thanks for the explanation, we'll try to use a different approach then. I must've misunderstood the documentation for this config option.

**Comment by traefiker** [comment]
Closed by #8072.
